src/types.h

```diff
@@ -438,41 +445,41 @@ inline int from_to(Move m) {
  return m & 0xFFF;
 }
 
-inline bool is_gating(Move m) {
-  return (m & (3 << 12)) && (rank_of(from_sq(m)) == RANK_1 || rank_of(from_sq(m)) == RANK_8) && ((m & (3 << 14)) == NORMAL || (m & (3 << 14)) == CASTLING || rank_of(from_sq(m)) == rank_of(to_sq(m)));
+inline MoveType type_of(Move m) {
+  return MoveType(m & CASTLING ? (m ^ (m >> 6)) & PROMOTION : m & ENPASSANT);
 }
 
-inline bool gating_on_castling_rook(Move m) {
-  return (m & (3 << 14)) == PROMOTION && (m & (3 << 12)) && rank_of(from_sq(m)) == rank_of(to_sq(m));
+inline bool is_normal(Move m) {
+  return !(m & (3 << 14));
 }
 
-inline MoveType type_of(Move m) {
-  if ((m & (3 << 14)) == PROMOTION2 && (m & (3 << 12)))
-      return PROMOTION;
-  if (gating_on_castling_rook(m))
-      return CASTLING;
-  return MoveType(m & (3 << 14));
+inline PieceType promotion_type(Move m) {
+  assert(type_of(m) == PROMOTION);
+  return PieceType((m >> 12) - (CASTLING >> 12));
```

Comment:
`(m ^ CASTLING) >> 12` (or `(m ^ CASTLING_OR_PROMOTION) >> 12`) might be simpler.